fix: improve required errors for validation levels

tjallingt opened the pull request:
Remove the `requiredVerified` and `requiredPublish` options from the JSON Schema, the checks are moved to after AJV has checked the schema. This allows us to improve the error message to mention that the error is _because_ of the validation level.

RobinBol commented on lib/App/index.js:
```diff
@@ -143,6 +120,16 @@ class App {
     }
 
+    if (level >= VALIDATION_LEVELS.VERIFIED) {
+      if (appJson.platforms === undefined) {
+        throw new Error('The property `platforms` is required in order to publish a verified app.');
+      }
+
+      if (appJson.support === undefined) {
+        throw new Error('The property `support` is required in order to publish a verified app.');
+      }
```
This doesn't verify that `platforms` is an array and `support` a valid support string, not sure if that used to be the case though.

tjallingt replied:
The schema takes care of that, the schema just doesn't make sure the property actually exists.

RobinBol commented on lib/App/index.js:
```diff
@@ -152,8 +139,10 @@ class App {
-          console.warn('Warning: using the homey:manager:api permission will require a more thorough review. It may take longer than usual to review your app.');
+        if (permission === 'homey:manager:api') {
+          if (level >= VALIDATION_LEVELS.PUBLISH) {
+            console.warn('Warning: using the homey:manager:api permission will require a more thorough review. It may take longer than usual to review your app.');
```
Maybe this should use the `Log` method?

tjallingt replied:
Maybe, don't know if `node-homey-lib` as such a method. Code was already like this so seems fine to keep it like this.

RobinBol replied:
It does, might be more noticible using colors, but not an issue really: https://github.com/athombv/node-homey/search?q=Log

RobinBol commented on lib/App/index.js:
```diff
@@ -277,10 +268,20 @@ class App {
         }
 
+        if (level >= VALIDATION_LEVELS.VERIFIED) {
+          if (driver.platforms === undefined) {
+            throw new Error(`drivers.${driver.id}: property \`platforms\` is required in order to publish a verified app.`);
+          }
+
+          if (driver.connectivity === undefined) {
+            throw new Error(`drivers.${driver.id}: property \`connectivity\` is required in order to publish a verified app.`);
```
This doesn't verify the value of these properties

tjallingt replied:
The schema takes care of that, the schema just doesn't make sure the property actually exists.

RobinBol commented on lib/App/index.js:
```diff
@@ -580,36 +578,43 @@ class App {
+    if (level >= VALIDATION_LEVELS.VERIFIED) {
+      for (const argument of filteredCardArgs) {
+        if (argument.title === undefined) {
+          throw new Error(`${errorPath}: property \`title\` for is required for arguments in order to publish a verified app.`);
```
```suggestion
          throw new Error(`${errorPath}: property \`title\` is required for arguments in order to publish a verified app.`);
```

RobinBol commented on lib/App/index.js:
```diff
@@ -580,36 +578,43 @@ class App {
     }
 
+    if (typeof card.titleFormatted === 'undefined') {
+      console.warn(`Warning: ${errorPath}.titleFormatted is missing. Specifying a Flow card's formatted title will be required in the future.`);
```
`Log` perhaps?

RobinBol commented on package.json:
```diff
@@ -28,7 +28,7 @@
   "config": {
     "npmPublishTagProduction": "latest",
     "npmPublishTagTesting": "beta",
-    "nodeVersion": "8"
+    "nodeVersion": "12"
   },
```
I think this config object isn't actually used anymore

tjallingt replied:
Yea don't think so either, I can remove and add the `"engines"` field if we want that

WeeJeWel commented on helpers/index.js:
```diff
@@ -1,28 +1,18 @@
-
 'use strict';
 
-try {
```
Why has this been changed? It's for browser & react environments.

WeeJeWel commented on lib/Signal/index.js:
```diff
@@ -20,26 +20,13 @@ class Signal {
     this._check = this._check.bind(this);
   }
 
-  debug(...args) {
```
Why has this been removed?

WeeJeWel commented on lib/App/index.js:
```diff
@@ -3,35 +3,42 @@
+  extname,
+  basename,
+  dirname,
+} = require('path');
```
Please use `path.join()` in code, otherwise it's hard to trace where a function comes from.

WeeJeWel commented on lib/App/index.js:
```diff
@@ -3,35 +3,42 @@
-  'publish',
-  'verified',
-];
+const VALIDATION_LEVELS = {
```
It's not per se implied that levels follow up on each other. It's better to manually specify them for every check, leaving room for future levels with different requirements.

WeeJeWel commented on lib/App/index.js:
```diff
@@ -3,35 +3,42 @@
+const lstatAsync = util.promisify(fs.lstat);
+const imageSizeAsync = util.promisify(imageSize);
+
 const {
```
Ideally we make a class `Util` with static method `readBytesFromFile`. It's more consistent with our other repositories.

WeeJeWel commented:
@tjallingt I think now is a great moment to create tests that validates a spectrum of apps, where some are on purpose invalid.

tjallingt commented:
Cleaned up the changes to solve the comments, reverted the change to treat level as a number.
I think adding tests is a good idea, I'll make another PR for that.

WeeJeWel commented:
👍🏻  I'll wait with merging this until that PR is ready.

tjallingt commented:
The tests are in https://github.com/athombv/node-homey-lib/pull/330
I rebased this PR on that branch so it is easier to review the changes to the tests (just error message wording)

jeroenwienk commented on lib/App/index.js:
```diff
@@ -583,33 +595,40 @@ class App {
 
+    if (levelVerified) {
+      for (const argument of filteredCardArgs) {
+        if (argument.title === undefined) {
```
Why are these checks for `undefined` and not also `null`? Bit odd that someone would set a property to `null` but still possible.

tjallingt replied:
AJV will error if any of these are set to `null` , but `undefined` is allowed when a property is not required.
